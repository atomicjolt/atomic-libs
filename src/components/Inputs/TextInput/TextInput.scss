@use "../../mixins";
@use '../common.scss';

.aje-input {
  input[type="text"],
  input[type="password"],
  input[type="email"],
  input[type="tel"] {
    --input-width-sm: 100px;
    --input-width-md: 200px;
    --input-width-lg: 300px;
  }
}

.aje-input--floating {
  @extend .aje-input;
  --input-height: 48px;
  --input-padding-horiz: 1em;
  --input-bg-clr: var(--neutral100);
  --input-transition: 100ms ease;

  --floating-font-size: 1.6rem;
  --floating-label-clr: var(--text-clr-alt);

  .aje-input__container {
    position: relative;
  }

  label {
    position: absolute;
    top: 50%;
    left: calc(1em - 4px);
    transform: translateY(-50%);
    background-color: var(--input-bg-clr);
    color: var(--floating-label-clr);
    text-align: center;
    transition: 200ms ease;
    @include mixins.regular;
    font-size: var(--floating-font-size);
    pointer-events: none;
    padding: 1px 4px;
  }

  input:not(:placeholder-shown) ~ label,
  input:focus ~ label {
    top: 0px;
    left: 1em;
    --floating-font-size: 1.3rem;
    --input-bg-clr: var(--neutral50);
    --floating-label-clr: var(--text-clr);
  }

  input:not(:placeholder-shown) {
    box-shadow: 0 0 0 1px var(--input-border-color);
    --input-bg-clr: var(--neutral50);
  }

  input:focus {
    --input-border-clr: var(--primary700);
    --input-bg-clr: var(--neutral50);
    box-shadow: 0 0 0 1px var(--input-border-color);
    outline: none;
  }
  input:hover {
    --input-border-color: var(--neutral500);
    box-shadow: 0 0 0 1px var(--input-border-color);
  }
  input:focus:hover {
    --input-border-color: var(--primary800);
    box-shadow: 0 0 0 1px var(--input-border-color);
  }
  input:focus ~ label {
    --floating-label-clr: var(--text-clr);
  }

  &.has-error {
    --input-border-color: var(--error700);

    input:not(:placeholder-shown) ~ label,
    input:focus ~ label {
      --floating-label-clr: var(--error800);
    }
  }
}
