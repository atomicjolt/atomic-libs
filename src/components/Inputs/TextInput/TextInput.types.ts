import { HasChildren, HasVariant, InputProps } from "../../../types";

export type Variants = "default" | "floating";

export interface TextInputProps
  extends InputProps<string>,
    HasVariant<Variants> {
  /** Other types like date, time, and number have their own component  */
  type?: "text" | "email" | "tel" | "password";
}

export interface TextInputVariantProps extends HasChildren {
  message?: string;
  hideLabel?: boolean;
  label: string;
  inputId: string;
}
