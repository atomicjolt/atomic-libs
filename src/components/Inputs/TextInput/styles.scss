@use "../../defines" as *;

.aj-input {
  input[type="text"],
  input[type="password"],
  input[type="email"],
  input[type="tel"] {
    --input-width-sm: 100px;
    --input-width-md: 200px;
    --input-width-lg: 300px;
  }
}
