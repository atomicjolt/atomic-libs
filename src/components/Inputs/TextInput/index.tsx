import React from 'react'
import '../../general.scss'
import '../common.scss'
import './styles.scss'

export interface Props {
  type?: 'text' | 'password' | 'email' | 'tel';
  /** Must include a label. Labels are always Sentence case. */
  label: string;
  /** Error text should be descriptive and explicit in meaning. */
  error?: string;
  /** For additional information (ex. date format mm/dd/yy) */
  message?: string;
  /** Placeholders aren't common, you should use the message instead for most placeholders so the user doesn't lose the information after entering text in the input. */
  placeholder?: string;
  /** Only use in very specific circumstances. This hides the label from view, but still allows screen readers to read the label. (A search input might have 'Search...' as a placeholder and no room for a full label so you can hide it.) */
  hideLabel?: boolean;
  /** The input size should reflect the expected size of its content. */
  size?: 'small' | 'medium' | 'large' | 'auto' | 'full';
  value?: string;
  readonly?: boolean;
  disabled?: boolean;
  required?: boolean;
}

/** Text Input Component */
export default function TextInput({ 
    type = 'text',
    size = 'medium', 
    label, 
    error, 
    message, 
    placeholder, 
    hideLabel = false,
    value,
    readonly = false,
    disabled = false,
    required = false,
  }: Props) {

  const inputID = 'textInput';
  const errorID = 'errorText';
  /* Add a space before the added class rather than inside the className attr on the tag. Looks cleaner. */
  let errorClass = error ? ' has-error' : '';
  let disabledClass = disabled ? ' is-disabled' : '';
  let hiddenClass = hideLabel ? ' aj-hidden' : '';

  return (
    <div className={`aj-input is-${size}${errorClass}${disabledClass}`}>
      <label className={`aj-label${hiddenClass}`} htmlFor={inputID}>
        {label}
        {message ?
          <p className="aj-label--message">{message}</p>
          : null
        }
      </label>
      <input
        id={inputID}
        aria-describedby={error ? errorID : ""}
        type={type}
        placeholder={placeholder}
        value={value}
        readOnly={readonly}
        disabled={disabled}
        required={required}
      />
      {error ?
        <p id={errorID} className="aj-label--error">{error}</p>
        : null
      }
    </div>
  )
}
