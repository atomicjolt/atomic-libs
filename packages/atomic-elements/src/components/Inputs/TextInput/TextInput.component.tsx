import React from "react";
import { AriaTextFieldProps } from "@react-aria/textfield";
import classNames from "classnames";
import { useVariantClass } from "../../../hooks/variants";
import { AriaProps, FieldInputProps, HasVariant } from "../../../types";
import { StyledTextField } from "../Inputs.styles";
import { FieldInput } from "../../Fields";
import FloatingFieldInputWrapper from "../../Internal/FloatingFieldInputWrapper";

type Variants = "default" | "floating";

export interface TextInputProps
  extends AriaProps<AriaTextFieldProps>,
    FieldInputProps,
    HasVariant<Variants> {}

/** TextInput component. */
export const TextInput = React.forwardRef(function TextInput(
  props: TextInputProps,
  ref: React.Ref<HTMLInputElement>
) {
  const {
    type = "text",
    size = "medium",
    variant = "default",
    className,
    label,
    message,
    error,
    ...rest
  } = props;

  const variantClassName = useVariantClass("aje-input", variant);

  return (
    <StyledTextField
      type={type}
      size={size}
      className={classNames(variantClassName, className)}
      data-float={(Boolean(props.value) && variant === "floating") || undefined}
      {...rest}
    >
      <FloatingFieldInputWrapper
        floating={variant === "floating"}
        label={label}
        message={message}
        error={error}
      >
        <FieldInput ref={ref} />
      </FloatingFieldInputWrapper>
    </StyledTextField>
  );
});

export default TextInput;
