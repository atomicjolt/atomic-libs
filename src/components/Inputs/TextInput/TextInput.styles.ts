import styled from "styled-components";
import mixins from "../../../styles/mixins";
import { InputWrapper } from "../Inputs.styles";

export const TextInputWrapper = styled(InputWrapper)`
  input[type="text"],
  input[type="password"],
  input[type="email"],
  input[type="tel"] {
    --input-width-sm: 100px;
    --input-width-md: 200px;
    --input-width-lg: 300px;
  }

  &.aje-input--floating {
    --input-height: 48px;
    --input-padding-horiz: 1em;
    --input-bg-clr: var(--neutral100);
    --input-transition: 100ms ease;

    --floating-font-size: 1.6rem;
    --floating-label-clr: var(--text-clr-alt);

    label {
      ${mixins.Regular}
      z-index: 2;
      position: absolute;
      top: 50%;
      left: calc(1em - 4px);
      transform: translateY(-50%);
      background-color: var(--input-bg-clr);
      color: var(--floating-label-clr);
      text-align: center;
      transition: 200ms ease;
      font-size: var(--floating-font-size);
      pointer-events: none;
      padding: 1px 4px;
    }

    input:focus {
      --input-border-clr: var(--primary700);
      --input-bg-clr: var(--neutral50);
      box-shadow: 0 0 0 1px var(--input-border-color);
      outline: none;
    }

    &.has-value {
      --input-bg-clr: var(--neutral50);
    }

    input:focus ~ label,
    &.has-value label {
      top: 0px;
      left: 1em;
      --floating-font-size: 1.3rem;
      --input-bg-clr: var(--neutral50);
      --floating-label-clr: var(--text-clr);
    }

    input:hover {
      --input-border-color: var(--neutral500);
      box-shadow: 0 0 0 1px var(--input-border-color);
    }

    &.has-error {
      --input-border-color: var(--error700);

      input:not(:placeholder-shown) ~ label,
      input:focus ~ label {
        --floating-label-clr: var(--error800);
      }
    }
  }
`;

export const FloatingInputWrapper = styled.div`
  position: relative;
`;
