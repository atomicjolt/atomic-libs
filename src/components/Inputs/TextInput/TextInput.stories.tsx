import { ComponentMeta, ComponentStory } from "@storybook/react";
import TextInput, { Props } from ".";

export default {
title: "Inputs/TextInput",
component: TextInput
} as ComponentMeta<typeof TextInput>

const Template: ComponentStory<typeof TextInput> = (args: Props) => (
  <TextInput {...args} />
);

export const Default = Template.bind({});
Default.args = {
  label: "Text input label",
}

// Additional stories go here