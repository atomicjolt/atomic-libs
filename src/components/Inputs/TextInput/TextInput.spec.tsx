import React from "react";
import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import TextInput from ".";

describe("matches snapshot", () => {
  const shared = {
    label: "Text Label",
    message: "message",
    error: "some error",
    placeholder: "placeholder",
    value: "content",
    onChange: (value: string) => {},
  };

  test("default variant", () => {
    const result = render(<TextInput {...shared} />);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("floating variant", () => {
    const result = render(<TextInput variant="floating" {...shared} />);
    expect(result.asFragment()).toMatchSnapshot();
  });
});
