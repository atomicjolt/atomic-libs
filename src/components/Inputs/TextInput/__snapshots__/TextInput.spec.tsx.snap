// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot > default variant 1`] = `
<DocumentFragment>
  <div
    class="sc-gKAaef iJKtbz aje-input__text sc-bqGGcD hCFqor aje-input--default is-medium"
  >
    <label
      class="sc-bdnyFh KCKAU"
      for="react-aria-:r0:"
      id="react-aria-:r1:"
    >
      Text Label
    </label>
    <p
      class="sc-gtsqUy eWkPNm"
      id="react-aria-:r3:"
    >
      message
    </p>
    <input
      aria-describedby="react-aria-:r3:"
      aria-labelledby="react-aria-:r1:"
      class="sc-hKFymg iJvJd"
      id="react-aria-:r0:"
      placeholder="placeholder"
      type="text"
      value="content"
    />
  </div>
</DocumentFragment>
`;

exports[`matches snapshot > floating variant 1`] = `
<DocumentFragment>
  <div
    class="sc-gKAaef iJKtbz aje-input__text sc-bqGGcD hCFqor aje-input--floating is-medium"
    data-float="true"
  >
    <div
      class="sc-bkbjWr bivRPJ"
    >
      <input
        aria-describedby="react-aria-:r8:"
        aria-labelledby="react-aria-:r6:"
        class="sc-hKFymg iJvJd"
        data-float="true"
        id="react-aria-:r5:"
        placeholder="placeholder"
        type="text"
        value="content"
      />
      <label
        class="sc-bdnyFh KCKAU"
        for="react-aria-:r5:"
        id="react-aria-:r6:"
      >
        Text Label
      </label>
    </div>
    <p
      class="sc-gtsqUy eWkPNm"
      id="react-aria-:r8:"
    >
      message
    </p>
  </div>
</DocumentFragment>
`;
